File: apps/barrel_vectordb/src/barrel_vectordb_distance.erl
%%%-------------------------------------------------------------------
%%% @doc Vector Distance Computation Module
%%%
%%% Provides SIMD-accelerated distance computations via NIF.
%%% Falls back to pure Erlang implementation if NIF is not available.
%%%
%%% Supports:
%%% - Cosine distance
%%% - Euclidean distance
%%% - Dot product
%%%
%%% Vectors can be represented as:
%%% - Lists of floats (Erlang native, slower)
%%% - Binaries of 32-bit floats (NIF optimized, faster)
%%%
%%% @end
%%%-------------------------------------------------------------------
-module(barrel_vectordb_distance).

-export([
    %% Distance functions (work with both list and binary vectors)
    dot_product/2,
    cosine_distance/2,
    cosine_distance_normalized/2,
    euclidean_distance/2,

    %% Conversion functions
    to_binary/1,
    from_binary/1,

    %% Info
    simd_info/0,
    nif_available/0
]).


%% ============================================================
%% Distance functions
%% ============================================================

%% @doc Compute dot product of two vectors
-spec dot_product(vector(), vector()) -> float().
dot_product(Vec1, Vec2) when is_list(Vec1), is_list(Vec2) ->
    dot_product_list(Vec1, Vec2, 0.0);
dot_product(Bin1, Bin2) when is_binary(Bin1), is_binary(Bin2) ->
    nif_dot_product(Bin1, Bin2);
dot_product(Vec1, Vec2) when is_list(Vec1), is_binary(Vec2) ->
    dot_product(to_binary(Vec1), Vec2);
dot_product(Vec1, Vec2) when is_binary(Vec1), is_list(Vec2) ->
    dot_product(Vec1, to_binary(Vec2)).

%% @doc Compute cosine distance: 1 - cos(a, b)
-spec cosine_distance(vector(), vector()) -> float().
cosine_distance(Vec1, Vec2) when is_list(Vec1), is_list(Vec2) ->
    cosine_distance_list(Vec1, Vec2);
cosine_distance(Bin1, Bin2) when is_binary(Bin1), is_binary(Bin2) ->
    nif_cosine_distance(Bin1, Bin2);
cosine_distance(Vec1, Vec2) when is_list(Vec1), is_binary(Vec2) ->
    cosine_distance(to_binary(Vec1), Vec2);
cosine_distance(Vec1, Vec2) when is_binary(Vec1), is_list(Vec2) ->
    cosine_distance(Vec1, to_binary(Vec2)).

%% @doc Compute cosine distance for pre-normalized vectors (faster)
-spec cosine_distance_normalized(vector(), vector()) -> float().
cosine_distance_normalized(Vec1, Vec2) when is_list(Vec1), is_list(Vec2) ->
    1.0 - dot_product_list(Vec1, Vec2, 0.0);
cosine_distance_normalized(Bin1, Bin2) when is_binary(Bin1), is_binary(Bin2) ->
    nif_cosine_distance_normalized(Bin1, Bin2);
cosine_distance_normalized(Vec1, Vec2) when is_list(Vec1), is_binary(Vec2) ->
    cosine_distance_normalized(to_binary(Vec1), Vec2);
cosine_distance_normalized(Vec1, Vec2) when is_binary(Vec1), is_list(Vec2) ->
    cosine_distance_normalized(Vec1, to_binary(Vec2)).

%% @doc Compute Euclidean distance
-spec euclidean_distance(vector(), vector()) -> float().
euclidean_distance(Vec1, Vec2) when is_list(Vec1), is_list(Vec2) ->
    euclidean_distance_list(Vec1, Vec2);
euclidean_distance(Bin1, Bin2) when is_binary(Bin1), is_binary(Bin2) ->
    nif_euclidean_distance(Bin1, Bin2);
euclidean_distance(Vec1, Vec2) when is_list(Vec1), is_binary(Vec2) ->
    euclidean_distance(to_binary(Vec1), Vec2);
euclidean_distance(Vec1, Vec2) when is_binary(Vec1), is_list(Vec2) ->
    euclidean_distance(Vec1, to_binary(Vec2)).

%% ============================================================
%% Conversion functions
%% ============================================================

%% @doc Convert list of floats to binary vector (32-bit floats)
-spec to_binary([float()]) -> binary().
to_binary(List) when is_list(List) ->
    nif_to_binary(List).

%% @doc Convert binary vector to list of floats
-spec from_binary(binary()) -> [float()].
from_binary(Bin) when is_binary(Bin) ->
    nif_from_binary(Bin).

%% ============================================================
%% Info functions
%% ============================================================

%% @doc Return SIMD backend information
-spec simd_info() -> #{backend => avx2 | neon | scalar | erlang}.
simd_info() ->
    nif_simd_info().

%% @doc Check if NIF is available
-spec nif_available() -> boolean().
nif_available() ->
    try
        _ = simd_info(),
        true
    catch
        error:undef -> false
    end.

%% ============================================================
%% NIF calls (via unified barrel_vectordb_nif module)
%% ============================================================

nif_dot_product(Bin1, Bin2) ->
    try
        barrel_vectordb_nif:dot_product(Bin1, Bin2)
    catch
        error:undef ->
            dot_product_list(binary_to_list_fallback(Bin1), binary_to_list_fallback(Bin2), 0.0)
    end.

nif_cosine_distance(Bin1, Bin2) ->
    try
        barrel_vectordb_nif:cosine_distance(Bin1, Bin2)
    catch
        error:undef ->
            cosine_distance_list(binary_to_list_fallback(Bin1), binary_to_list_fallback(Bin2))
    end.

nif_cosine_distance_normalized(Bin1, Bin2) ->
    try
        barrel_vectordb_nif:cosine_distance_normalized(Bin1, Bin2)
    catch
        error:undef ->
            1.0 - nif_dot_product(Bin1, Bin2)
    end.

nif_euclidean_distance(Bin1, Bin2) ->
    try
        barrel_vectordb_nif:euclidean_distance(Bin1, Bin2)
    catch
        error:undef ->
            euclidean_distance_list(binary_to_list_fallback(Bin1), binary_to_list_fallback(Bin2))
    end.

nif_to_binary(List) ->
    try
        barrel_vectordb_nif:to_binary(List)
    catch
        error:undef ->
            << <<F:32/float-little>> || F <- List >>
    end.

nif_from_binary(Bin) ->
    try
        barrel_vectordb_nif:from_binary(Bin)
    catch
        error:undef ->
            binary_to_list_fallback(Bin)
    end.

nif_simd_info() ->
    try
        #{backend => barrel_vectordb_nif:simd_info()}
    catch
        error:undef ->
            #{backend => erlang}
    end.

%% ============================================================
%% Erlang fallback implementations
%% ============================================================

-type vector() :: [float()] | binary().

binary_to_list_fallback(<<>>) -> [];
binary_to_list_fallback(<<F:32/float-little, Rest/binary>>) ->
    [F | binary_to_list_fallback(Rest)].

dot_product_list([], [], Acc) -> Acc;
dot_product_list([A | R1], [B | R2], Acc) ->
    dot_product_list(R1, R2, Acc + A * B).

cosine_distance_list(Vec1, Vec2) ->
    Dot = dot_product_list(Vec1, Vec2, 0.0),
    Norm1 = math:sqrt(dot_product_list(Vec1, Vec1, 0.0)),
    Norm2 = math:sqrt(dot_product_list(Vec2, Vec2, 0.0)),
    Denom = Norm1 * Norm2,
    case Denom < 1.0e-10 of
        true -> 1.0;
        false -> 1.0 - (Dot / Denom)
    end.

euclidean_distance_list(Vec1, Vec2) ->
    SqDist = euclidean_sq_list(Vec1, Vec2, 0.0),
    math:sqrt(SqDist).

euclidean_sq_list([], [], Acc) -> Acc;
euclidean_sq_list([A | R1], [B | R2], Acc) ->
    Diff = A - B,
    euclidean_sq_list(R1, R2, Acc + Diff * Diff).

File: apps/barrel_vectordb/c_src/barrel_vectordb_distance.c
/**
 * barrel_vectordb_distance.c
 *
 * NIF implementation for SIMD-accelerated vector distance computation.
 * Supports AVX2 on x86_64 and NEON on ARM (if available).
 */

#include "barrel_vectordb_nif_common.h"

/* ============================================================
 * Fallback (scalar) implementations
 * Used when SIMD is not available, or for norm calculation
 * ============================================================ */

#if !defined(USE_AVX2) && !defined(USE_NEON)
static double dot_product_scalar(const float* a, const float* b, int n) {
    double sum = 0.0;
    for (int i = 0; i < n; i++) {
        sum += (double)a[i] * (double)b[i];
    }
    return sum;
}

static double euclidean_distance_sq_scalar(const float* a, const float* b, int n) {
    double sum = 0.0;
    for (int i = 0; i < n; i++) {
        double diff = (double)a[i] - (double)b[i];
        sum += diff * diff;
    }
    return sum;
}
#endif

static double norm_scalar(const float* a, int n) {
    double sum = 0.0;
    for (int i = 0; i < n; i++) {
        sum += (double)a[i] * (double)a[i];
    }
    return sqrt(sum);
}

/* ============================================================
 * SIMD implementations
 * ============================================================ */

#ifdef USE_AVX2
static double dot_product_avx2(const float* a, const float* b, int n) {
    __m256 sum = _mm256_setzero_ps();
    int i = 0;

    /* Process 8 floats at a time */
    for (; i + 7 < n; i += 8) {
        __m256 va = _mm256_loadu_ps(a + i);
        __m256 vb = _mm256_loadu_ps(b + i);
        sum = _mm256_fmadd_ps(va, vb, sum);
    }

    /* Horizontal sum */
    __m128 hi = _mm256_extractf128_ps(sum, 1);
    __m128 lo = _mm256_castps256_ps128(sum);
    __m128 sum128 = _mm_add_ps(hi, lo);
    sum128 = _mm_hadd_ps(sum128, sum128);
    sum128 = _mm_hadd_ps(sum128, sum128);

    double result = _mm_cvtss_f32(sum128);

    /* Handle remaining elements */
    for (; i < n; i++) {
        result += (double)a[i] * (double)b[i];
    }

    return result;
}

static double euclidean_distance_sq_avx2(const float* a, const float* b, int n) {
    __m256 sum = _mm256_setzero_ps();
    int i = 0;

    for (; i + 7 < n; i += 8) {
        __m256 va = _mm256_loadu_ps(a + i);
        __m256 vb = _mm256_loadu_ps(b + i);
        __m256 diff = _mm256_sub_ps(va, vb);
        sum = _mm256_fmadd_ps(diff, diff, sum);
    }

    __m128 hi = _mm256_extractf128_ps(sum, 1);
    __m128 lo = _mm256_castps256_ps128(sum);
    __m128 sum128 = _mm_add_ps(hi, lo);
    sum128 = _mm_hadd_ps(sum128, sum128);
    sum128 = _mm_hadd_ps(sum128, sum128);

    double result = _mm_cvtss_f32(sum128);

    for (; i < n; i++) {
        double diff = (double)a[i] - (double)b[i];
        result += diff * diff;
    }

    return result;
}
#endif

#ifdef USE_NEON
static double dot_product_neon(const float* a, const float* b, int n) {
    float32x4_t sum = vdupq_n_f32(0.0f);
    int i = 0;

    for (; i + 3 < n; i += 4) {
        float32x4_t va = vld1q_f32(a + i);
        float32x4_t vb = vld1q_f32(b + i);
        sum = vmlaq_f32(sum, va, vb);
    }

    float result = vaddvq_f32(sum);

    for (; i < n; i++) {
        result += a[i] * b[i];
    }

    return (double)result;
}

static double euclidean_distance_sq_neon(const float* a, const float* b, int n) {
    float32x4_t sum = vdupq_n_f32(0.0f);
    int i = 0;

    for (; i + 3 < n; i += 4) {
        float32x4_t va = vld1q_f32(a + i);
        float32x4_t vb = vld1q_f32(b + i);
        float32x4_t diff = vsubq_f32(va, vb);
        sum = vmlaq_f32(sum, diff, diff);
    }

    float result = vaddvq_f32(sum);

    for (; i < n; i++) {
        float diff = a[i] - b[i];
        result += diff * diff;
    }

    return (double)result;
}
#endif

/* ============================================================
 * Dispatch functions (choose best implementation)
 * ============================================================ */

static double do_dot_product(const float* a, const float* b, int n) {
#ifdef USE_AVX2
    return dot_product_avx2(a, b, n);
#elif defined(USE_NEON)
    return dot_product_neon(a, b, n);
#else
    return dot_product_scalar(a, b, n);
#endif
}

static double do_euclidean_distance_sq(const float* a, const float* b, int n) {
#ifdef USE_AVX2
    return euclidean_distance_sq_avx2(a, b, n);
#elif defined(USE_NEON)
    return euclidean_distance_sq_neon(a, b, n);
#else
    return euclidean_distance_sq_scalar(a, b, n);
#endif
}

/* ============================================================
 * NIF functions
 * ============================================================ */

/**
 * dot_product(Binary1, Binary2) -> float()
 */
ERL_NIF_TERM nif_dot_product(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
    (void)argc;
    ErlNifBinary bin1, bin2;

    if (!enif_inspect_binary(env, argv[0], &bin1) ||
        !enif_inspect_binary(env, argv[1], &bin2)) {
        return enif_make_badarg(env);
    }

    if (bin1.size != bin2.size || bin1.size % sizeof(float) != 0) {
        return enif_make_badarg(env);
    }

    int n = bin1.size / sizeof(float);
    const float* a = (const float*)bin1.data;
    const float* b = (const float*)bin2.data;

    double result = do_dot_product(a, b, n);

    return enif_make_double(env, result);
}

/**
 * cosine_distance(Binary1, Binary2) -> float()
 */
ERL_NIF_TERM nif_cosine_distance(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
    (void)argc;
    ErlNifBinary bin1, bin2;

    if (!enif_inspect_binary(env, argv[0], &bin1) ||
        !enif_inspect_binary(env, argv[1], &bin2)) {
        return enif_make_badarg(env);
    }

    if (bin1.size != bin2.size || bin1.size % sizeof(float) != 0) {
        return enif_make_badarg(env);
    }

    int n = bin1.size / sizeof(float);
    const float* a = (const float*)bin1.data;
    const float* b = (const float*)bin2.data;

    double dot = do_dot_product(a, b, n);
    double norm_a = norm_scalar(a, n);
    double norm_b = norm_scalar(b, n);

    double denom = norm_a * norm_b;
    if (denom < 1.0e-10) {
        return enif_make_double(env, 1.0);
    }

    double cosine = dot / denom;
    double distance = 1.0 - cosine;

    return enif_make_double(env, distance);
}

/**
 * cosine_distance_normalized(Binary1, Binary2) -> float()
 */
ERL_NIF_TERM nif_cosine_distance_normalized(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
    (void)argc;
    ErlNifBinary bin1, bin2;

    if (!enif_inspect_binary(env, argv[0], &bin1) ||
        !enif_inspect_binary(env, argv[1], &bin2)) {
        return enif_make_badarg(env);
    }

    if (bin1.size != bin2.size || bin1.size % sizeof(float) != 0) {
        return enif_make_badarg(env);
    }

    int n = bin1.size / sizeof(float);
    const float* a = (const float*)bin1.data;
    const float* b = (const float*)bin2.data;

    double dot = do_dot_product(a, b, n);
    double distance = 1.0 - dot;

    return enif_make_double(env, distance);
}

/**
 * euclidean_distance(Binary1, Binary2) -> float()
 */
ERL_NIF_TERM nif_euclidean_distance(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
    (void)argc;
    ErlNifBinary bin1, bin2;

    if (!enif_inspect_binary(env, argv[0], &bin1) ||
        !enif_inspect_binary(env, argv[1], &bin2)) {
        return enif_make_badarg(env);
    }

    if (bin1.size != bin2.size || bin1.size % sizeof(float) != 0) {
        return enif_make_badarg(env);
    }

    int n = bin1.size / sizeof(float);
    const float* a = (const float*)bin1.data;
    const float* b = (const float*)bin2.data;

    double sq_dist = do_euclidean_distance_sq(a, b, n);
    double distance = sqrt(sq_dist);

    return enif_make_double(env, distance);
}

/**
 * to_binary(List) -> binary()
 */
ERL_NIF_TERM nif_to_binary(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
    (void)argc;
    unsigned int len;

    if (!enif_get_list_length(env, argv[0], &len)) {
        return enif_make_badarg(env);
    }

    ERL_NIF_TERM binary;
    float* data = (float*)enif_make_new_binary(env, len * sizeof(float), &binary);
    if (!data) {
        return enif_make_badarg(env);
    }

    ERL_NIF_TERM list = argv[0];
    ERL_NIF_TERM head, tail;
    unsigned int i = 0;

    while (enif_get_list_cell(env, list, &head, &tail)) {
        double val;
        if (!enif_get_double(env, head, &val)) {
            /* Try integer */
            long lval;
            if (!enif_get_long(env, head, &lval)) {
                return enif_make_badarg(env);
            }
            val = (double)lval;
        }
        data[i++] = (float)val;
        list = tail;
    }

    return binary;
}

/**
 * from_binary(Binary) -> [float()]
 */
ERL_NIF_TERM nif_from_binary(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]) {
    (void)argc;
    ErlNifBinary bin;

    if (!enif_inspect_binary(env, argv[0], &bin) || bin.size % sizeof(float) != 0) {
        return enif_make_badarg(env);
    }

    int n = bin.size / sizeof(float);
    const float* data = (const float*)bin.data;

    ERL_NIF_TERM* terms = enif_alloc(n * sizeof(ERL_NIF_TERM));
    if (!terms) {
        return enif_make_badarg(env);
    }

    for (int i = 0; i < n; i++) {
        terms[i] = enif_make_double(env, (double)data[i]);
    }

    ERL_NIF_TERM list = enif_make_list_from_array(env, terms, n);
    enif_free(terms);

    return list;
}
